### PR TITLE
Improve control over bindTemplate rendering

### LINKED ISSUE #26: [bindTemplate] be smarter about immediateRender
Currently, the `immediateRender` option for `bindTemplate` explicitly defines if a render should be done to fill an empty container, or if the developer already know that existing HTML is added by the server, and can delay the first execution.

Finding a good default is hard, and if the developer forgets to set it correctly, it might result in empty HTML – or wasteful execution.

Instead, we should be able to detect if there is already HTML present in the container that it is bound to. We can use this to indicate the default behaviour.

The option could then be changed to `forceRender` to indicate that it should always force an initial render as an override the the default behaviour.

Tasks:
- [ ] add test story
- [ ] implement feature
- [ ] update docs
- [ ] note as breaking change

### LINKED ISSUE #30: [bindTemplate] See if we can combine template + data into a render function that is watched
Currently the `bindTemplate` is receiving observable data and a template function as separate parameters (among other params).

The only reason to receive the data, is to watch it, and to pass it to the template function – that actually uses it.

An improvement here could be to require the template function to be a computed (having to return a template result), and make that implicitly watchable by the data that is used inside of it.

This might partially conflict with some other features that delay watching if there is already HTML present in the container. Which means the function should always be called once to set up tracking, but the result might need to be "ignored" the first time it's called.

### PR DESCRIPTION
**Pull request by ThaNarie**
* autodetect initial rendering
* change immediateRender as a required config, to an override from the default
* combine the data and template function to be more integrated
* update docs
* add additional stories

fixes #26, fixes #30

### CODE REVIEW

File: src/lib/bindings/applyBindings.ts [hunk 1]
```diff
@@ -143,28 +143,46 @@ export const applyBindings = (
...
-                instance.children.forEach((component) => component.setup());
-              }, 1);
+        watchEffect(() => {
+          if (ref?.element) {
+            // if neither of these are true, we should not do an initial render,
+            // but instead only "watch" the data in the onUpdate function
+            const shouldRenderUpdate = containerIsEmpty || forceImmediateRender || !initialRender;
+
+            // TODO: attach parent component for context
+            // pass along ohw the result of the update function is going to be used,
```
Comment (psimk) [inline src/lib/bindings/applyBindings.ts:167]:
```suggestion
            // pass along how the result of the update function is going to be used,
```

File: src/lib/bindings/applyBindings.ts [hunk 2]
```diff
@@ -143,28 +143,46 @@ export const applyBindings = (
...
+            // so the implementation can conditionally only invoke the watched observables, but
+            // omit the template rendering
+            const result = onUpdate(!shouldRenderUpdate);
+
+            if (shouldRenderUpdate) {
+              ref.element.innerHTML = result ? [].concat(result as any).join('') : '';
```
Comment (psimk) [inline src/lib/bindings/applyBindings.ts:173]: `result` can be `string | Array<string> | undefined`, right? So we only care to apply `[].concat` if `result` is an array:
```suggestion
              ref.element.innerHTML = Array.isArray(result) ? [].concat(result).join('') : result ?? '';
```

**Review by psimk — CHANGES_REQUESTED** [review]

File: src/lib/bindings/applyBindings.ts (hunk 2 above)
Comment (ThaNarie) [reply src/lib/bindings/applyBindings.ts:173]: The `concat` is a way to convert arrays and non-arrays to arrays, so you can join them whichever option is passed (TS just doesn't understand this).

If you do the `isArray` check, you don't need the `concat`, you can directly do the `join`.

Comment (psimk) [reply src/lib/bindings/applyBindings.ts:173]: So we can just omit `concat` then
```suggestion
              ref.element.innerHTML =  Array.isArray(result) ? result.join('') : result ?? '';
```

**Review by psimk — APPROVED** [review]

**Review by ThijsTyZ — APPROVED** [review]

**Pull request merged**